### DISCUSSION TITLE
Minimal example of using screen/multiply wanted

### DISCUSSION THREAD

**Discussion by ProgrammingLife** [Ogeon/palette#266] [General]
Could you please provide me with a little example of using screen/multiply blend modes to get new modified pixel?
My attempt is:
```
use palette::{ Hsl, IntoColor, Srgb };
let mut img1 = image::open( "data/img1.png" ).unwrap();
let mut img2 = image::open( "data/img2.png" ).unwrap();
let c1 = img1.get_pixel( 10, 10 );
let c2 = img2.get_pixel( 10, 10 );
let new_pixel: Hsl = Srgb::from( c2.0 ).into_format().into_color(); // <-- Here is the problem
// img1.put_pixel( 10, 10, new_pixel.into_format().into() );
```
I get following error message near "Srgb::from":
```
[rustc E0277] [E] the trait bound `palette::rgb::Rgb<palette::encoding::Srgb, _>: std::convert::From<[u8; 4]>` is not satisfied
the trait `std::convert::From<[u8; 4]>` is not implemented for `palette::rgb::Rgb<palette::encoding::Srgb, _>`
help: the following implementations were found:
        <palette::rgb::Rgb<S, T> as std::convert::From<(T, T, T)>>
        <palette::rgb::Rgb<S, u8> as std::convert::From<palette::Packed<C>>>
        <palette::rgb::Rgb<S, u8> as std::convert::From<u32>>
        <palette::rgb::Rgb<palette::encoding::Linear<palette::encoding::Srgb>, U> as std::convert::From<palette::rgb::Rgb<palette::encoding::Srgb, T>>>
        <palette::rgb::Rgb<palette::encoding::Srgb, U> as std::convert::From<palette::rgb::Rgb<palette::encoding::Linear<palette::encoding::Srgb>, T>>>
```

What do I do wrong? How to fix this issue?

# Env:
rust: *rustc 1.57.0 (f1edd0429 2021-11-29)*
os: *Linux Mint 19.1*
Cargo.toml:
```
[dependencies]
image = "0.23.14"
palette = "0.6"
```

**Comment by Ogeon** [comment]
Hello! Similar to your question about the hue related trait, if you are using the examples in the repo at the moment, they will be written for a future version of the library. The recommended way to convert from an array in 0.6.0 is to use the [Pixel](https://docs.rs/palette/0.6.0/palette/encoding/pixel/trait.Pixel.html) trait. After you have your `Srgba` color, you can follow the examples in the `blend` module: https://docs.rs/palette/0.6.0/palette/blend/index.html.

**Comment by ProgrammingLife** [comment]
Thanks for the reply! I don't understand how to convert it back to image::Pixel after having ```c```-variable here:
```
	let a = LinSrgba::new( 0.2, 0.5, 0.1, 0.8 );
	let b = LinSrgba::new( 0.6, 0.3, 0.5, 0.1 );
	let c = a.overlay( b );

	let new_pixel: Hsl = Srgba::from( &c ).into_format().into_color();
	let new_pixel: Hsl = Srgba::from_raw( &c ).into_format().into_color();

	img1.put_pixel( 10, 10, new_pixel );
	img1.put_pixel( 10, 10, image::Rgba( Srgba::from_color( c.into_format().into_color() ) ) ); // mistmatched types
```

```
#[derive(PartialEq, Debug, palette::Pixel)]
#[repr(C)]
struct MyRGBA {
    red: u8,
    green: u8,
    blue: u8,
    alpha: u8,
}
// ...
	let a = LinSrgba::new( 0.2, 0.5, 0.1, 0.8 );
	let b = LinSrgba::new( 0.6, 0.3, 0.5, 0.1 );
	let c = a.overlay( b );

	let new_pixel: Hsl = MyRGBA::from_raw( &c ).into_format().into_color();
```

I have no clue how should it work...

**Reply by Ogeon** [reply]
In the lingo of the Pixel trait, "raw" color data is arrays and slices. That's what you have in your image:

```rust
use palette::{ Hsl, IntoColor, Srgb, Pixel, Blend };
let mut img1 = image::open( "data/img1.png" ).unwrap();
let mut img2 = image::open( "data/img2.png" ).unwrap();
let c1 = img1.get_pixel( 10, 10 ); // c1.0 is an array
let c2 = img2.get_pixel( 10, 10 ); // c2.0 is an array
```

So to work with the colors c1 and c2 with Palette color types, you first need to convert from "raw" format:

```rust
let a = Srgba::from_raw( c1.0 ).into_format().into_linear();
let b = Srgba::from_raw( c2.0 ).into_format().into_linear();
```

Then we can overlay them:

```rust
let new_pixel = a.overlay( b ); // new_pixel is also LinSrgba
```

Finally we convert it back to non-linear, then u8 and then "raw":

```rust
img1.put_pixel( 10, 10, image::Rgba( Srgba::from_color( new_pixel ).into_format().into_raw() ) );
```

This is completely untested, but it may give you an idea of how it's meant to work. The whole thing with "raw" and the way the Pixel trait works is a bit funky, hence why I have rewritten it completely in the upcoming version. :)

**Reply by ProgrammingLife** [reply]
> ```rust
> use palette::{ Hsl, IntoColor, Srgb, Pixel, Blend };
> let mut img1 = image::open( "data/img1.png" ).unwrap();
> let c1 = img1.get_pixel( 10, 10 ); // c1.0 is an array
> let a = Srgba::from_raw( c1.0 ).into_format().into_linear();
> ```

Thanks for trying me to help but it doesn't work either:
```
[rustc E0599] [E] no function or associated item named `from_raw` found for struct `palette::Alpha<palette::rgb::Rgb<palette::encoding::Srgb, _>, _>` in the current scope
function or associated item not found in `palette::Alpha<palette::rgb::Rgb<palette::encoding::Srgb, _>, _>`
help: items from traits can only be used if the trait is in scope
```

When will be the upcoming version released?

**Reply by Ogeon** [reply]
You have probably not imported the `Pixel` trait, using `use palette::Pixel` somewhere, or like in the first part of the example. "help: items from traits can only be used if the trait is in scope" is a hint of that. Traits have to be imported for their methods to be callable.

As for the next version, I don't set release dates. I have a few more traits I want to fix up first. You can always use a git URL in you dependencies if you don't intend to publish your project on crates.io, but that comes with a higher risk of breaking changes whenever you update the Palette dependency. See the cargo documentation if you want to try that: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories
